discovery/config.go

```diff
@@ -37,6 +37,19 @@ type Config struct {
 	// This defaults to 1 minute.
 	ServerWatchDisabledInterval time.Duration
 
+	// ServerEvalFn is optional. It can be used to exclude servers based on
+	// custom criteria. If not nil, it is called after connecting to a server
+	// but prior to marking the server "current". When this returns false,
+	// the Watcher will skip the server.
+	//
+	// The State will be valid: The GRPCConn will be valid to use and
+	// DataplaneFeatures will be populated and the Address and Token (if
+	// applicable) will be set.
+	//
+	// This is called synchronously in the same goroutine as Watcher.Run(),
+	// so it should not block, or at least not for too long.
+	ServerEvalFn func(State) bool
```

Comment:
Nice, supporting arbitrary functions is much more flexible!

Filtering by features seems like it'd be pretty common though, so what do you think of adding a package-level function for it?

```Go
func SupportsDataplaneFeatures(names ...string) func(State) bool {
    return func(s State) bool {
        for _, name := range names {
            if !s.DataplaneFeatueres[name] {
                return false
            }
        }
        return true
    }
}

cfg := Config{
    ServerEvalFn: SupportsDataplaneFeatures("ENVOY_BOOTSTRAP_CONFIGURATION"),
}
```

Reply:
Yep, great suggestion! I added this in https://github.com/hashicorp/consul-server-connection-manager/pull/7/commits/2dca1ede0a7aff6e5e950bce27d4145f35b60895.

I saw the comment here about expose the protobuf enum directly: https://github.com/hashicorp/consul-dataplane/pull/7/files#diff-02b092271b5883db7a7fea12b029529c6859c6ceff872facc7b0e401041ff051R114-R116. I think that makes sense, as long as proto-public being a go module is our "forever" solution for protobuf code reuse.

Reply:
Awesome thanks!

I think the module is the solution we're stuck with until Go's protobuf implementation supports the same namespace in multiple Go packages. I'm not sure if/when that'll happen though, as the central registry seems to be required for `any` types.